util/LoginManager.php

```diff
@@ -151,20 +155,61 @@ public static function logout()
    * @return UserVO the User if it's logged in already, or a NULL
    * value otherwise.
    */
-  public static function isLogged($sid = NULL) {
-
+  public static function isLogged( $sid = NULL )
+  {
     if ($sid)
       session_id($sid);
 
     if (!isset($_SESSION))
       session_start();
 
     if (empty($_SESSION['user']))
-        return NULL;
+      return NULL;
 
     return $_SESSION['user'];
 
+  }
+
+  public static function isTokenActive( $token )
+  {
+    $oidc = self::setupOidcClient();
+    $intro_token = $oidc->introspectToken(
+      $token,
+      null,
+      ConfigurationParametersManager::getParameter('OIDC_CLIENT_ID'),
+      ConfigurationParametersManager::getParameter('JWT_SECRET')
+    );
+    return $intro_token->active;
+  }
+
+  public static function refreshAccessToken()
+  {
+    //tokens from session, stored originally on login
+    $current_refresh_token = $_SESSION['refresh_token'];
+    $current_access_token = $_SESSION['api_token'];
+
+    if ($current_access_token && $current_refresh_token) {
+      //introspect the access token; if it's good, move along; if not, go next to introspect the refresh token and if valid, do a refresh and get new access_token and new refresh token
+      if (!self::isTokenActive($current_access_token)) {
+        //now check the refresh token
+        if (self::isTokenActive($current_refresh_token)) {
+          //fyi - the refreshToken() method brings back an object
+          $refreshed_token = self::setupOidcClient()->refreshToken($current_refresh_token, false);
+          unset($_SESSION['api_token']);
+          $_SESSION['api_token'] = $refreshed_token->access_token;
+          $new_refresh_token = $refreshed_token->refresh_token;
+          unset($_SESSION['refresh_token']);
+          $_SESSION['refresh_token'] = $new_refresh_token;
+        } else {
+          //both access and refresh tokens have expired, so we need to authenticate again
+          self::setupOidcClient()->authenticate();
```

Comment:
I'm testing locally and it seems this part is not working properly, at least here once it gets in this point I start getting these errors in the console and it doesn't call the python API anymore:

```
Acess to XMLHttpRequest at 'https://xxxxx/realms/xxxx/protocol/openid-connect/auth?response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fweb%2Fservices%2FgetTaskTypes.php&client_id=phpreport-local&state=647a53a8fdea316458ad7fd2698bfd72&scope=openid&nonce=9e2e856bcaca6bdf4bc5f254086080b9&code_challenge=sGfFWzjYC5-vZKuSFRdc0PKlwqRd11IVCS3MmbFsgn4&code_challenge_method=S256' (redirected from 'http://localhost:8000/web/services/getTaskTypes.php?_dc=1691588764359') from origin 'http://localhost:8000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

GET https://xxxxx/realms/xxxx/protocol/openid-connect/auth?response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fweb%2Fservices%2FgetTaskTypes.php&client_id=phpreport-local&state=ca7395b6fc020b6ccbee1e5e6d292a32&scope=openid&nonce=1f7faadea6b4fcb7f7fb21e7c296908a&code_challenge=6BYzbcGmIRbfRc7XbwzcbUTpA5RptNTZV05JSJuyCO0&code_challenge_method=S256 net::ERR_FAILED
asyncR
```